subsys/socket_offload/providers/simplelink.c

```diff
@@ -0,0 +1,457 @@
+		}
+	} else if (sl_addr->sa_family == SL_AF_INET6) {
+		if (sl_addrlen == sizeof(SlSockAddrIn6_t)) {
+			struct sockaddr_in6 *z_sockaddr_in6 =
+				(struct sockaddr_in6 *)addr;
+			sl_addr_in6 = (SlSockAddrIn6_t *)sl_addr;
+
+			z_sockaddr_in6->sin6_family = AF_INET6;
+			z_sockaddr_in6->sin6_port = sl_addr_in6->sin6_port;
+			z_sockaddr_in6->sin6_scope_id =
+				(u8_t)sl_addr_in6->sin6_scope_id;
+			memcpy(z_sockaddr_in6->sin6_addr.s6_addr,
+			       sl_addr_in6->sin6_addr._S6_un._S6_u32,
+			       sizeof(z_sockaddr_in6->sin6_addr.s6_addr));
+			*addrlen = sizeof(struct sockaddr_in6);
+		} else {
+			*addrlen = sl_addrlen;
+		}
+	}
+}
+
+static int simplelink_accept(int sd, struct sockaddr *addr, socklen_t *addrlen)
+{
+	int retval;
+	SlSockAddr_t *sl_addr;
+	SlSockAddrIn_t sl_addr_in;
+	SlSockAddrIn6_t sl_addr_in6;
+	SlSocklen_t sl_addrlen;
+
+	if ((addrlen == NULL) || (addr == NULL)) {
+		retval = SL_RET_CODE_INVALID_INPUT;
+		goto exit;
+	}
+
+	/* Translate between Zephyr's and SimpleLink's sockaddr's: */
+	sl_addr = translate_z_to_sl_addrlen(*addrlen, &sl_addr_in, &sl_addr_in6,
+				  &sl_addrlen);
+	if (sl_addr == NULL) {
+		retval = SL_RET_CODE_INVALID_INPUT;
+		goto exit;
+	}
+
+	retval = sl_Accept(sd, sl_addr, &sl_addrlen);
+	if (retval < 0) {
+		goto exit;
+	}
+
+	/* Translate returned sl_addr into *addr and set *addrlen: */
+	translate_sl_to_z_addr(sl_addr, sl_addrlen, addr, addrlen);
+
+exit:
+	return _SlDrvSetErrno(retval);
+}
+
+static int simplelink_bind(int sd, const struct sockaddr *addr,
+			   socklen_t addrlen)
+{
+	int retval;
+	SlSockAddr_t *sl_addr;
+	SlSockAddrIn_t sl_addr_in;
+	SlSockAddrIn6_t sl_addr_in6;
+	SlSocklen_t sl_addrlen;
+
+	if (addr == NULL) {
+		retval = EISDIR;
+		goto exit;
+	}
+
+	/* Translate to sl_Bind() parameters: */
+	sl_addr = translate_z_to_sl_addrs(addr, addrlen, &sl_addr_in,
+					  &sl_addr_in6, &sl_addrlen);
+
+	if (sl_addr == NULL) {
+		retval = SL_RET_CODE_INVALID_INPUT;
+		goto exit;
+	}
+
+	retval = sl_Bind(sd, sl_addr, sl_addrlen);
+
+exit:
+	return _SlDrvSetErrno(retval);
+}
+
+static int simplelink_listen(int sd, int backlog)
+{
+	int retval;
+
+	retval = (int)sl_Listen(sd, backlog);
+
+	return _SlDrvSetErrno(retval);
+}
+
+static int simplelink_connect(int sd, const struct sockaddr *addr,
+			      socklen_t addrlen)
+{
+	int retval;
+	SlSockAddr_t *sl_addr;
+	SlSockAddrIn_t sl_addr_in;
+	SlSockAddrIn6_t sl_addr_in6;
+	SlSocklen_t sl_addrlen;
+
+	__ASSERT_NO_MSG(addr);
+
+	/* Translate to sl_Connect() parameters: */
+	sl_addr = translate_z_to_sl_addrs(addr, addrlen, &sl_addr_in,
+					  &sl_addr_in6, &sl_addrlen);
+
+	if (sl_addr == NULL) {
+		retval = SL_RET_CODE_INVALID_INPUT;
+		goto exit;
+	}
+
+	retval = sl_Connect(sd, sl_addr, sl_addrlen);
+
+exit:
+	return _SlDrvSetErrno(retval);
+}
+
+/* Excerpted from SimpleLink's socket.h:
+ * "Unsupported: these are only placeholders to not break BSD code.
+ *  Remove once Zephyr has POSIX socket options defined."
```

Comment:
I guess sentence about Zephyr is not part of the quote from SimpleLink's socket.h.

Reply:
Yup, thanks.